In [2]:
import pandas as pd

We're loading this dataset from a CSV and designating the default method to give index number to row

In [8]:
movie = pd.read_csv("C:\Python310\\imdbmovie.csv")

In [9]:
movie

Rank                    Title                     Genre  \
0       1  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1       2               Prometheus  Adventure,Mystery,Sci-Fi   
2       3                    Split           Horror,Thriller   
3       4                     Sing   Animation,Comedy,Family   
4       5            Suicide Squad  Action,Adventure,Fantasy   
..    ...                      ...                       ...   
995   996     Secret in Their Eyes       Crime,Drama,Mystery   
996   997          Hostel: Part II                    Horror   
997   998   Step Up 2: The Streets       Drama,Music,Romance   
998   999             Search Party          Adventure,Comedy   
999  1000               Nine Lives     Comedy,Family,Fantasy   

                                           Description              Director  \
0    A group of intergalactic criminals are forced ...            James Gunn   
1    Following clues to the origin of mankind, a te...          Ridley Scott   
2    Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3    In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4    A secret government agency recruits some of th...            David Ayer   
..                                                 ...                   ...   
995  A tight-knit team of rising investigators, alo...             Billy Ray   
996  Three American college students studying abroa...              Eli Roth   
997  Romantic sparks occur between two dance studen...            Jon M. Chu   
998  A pair of friends embark on a mission to reuni...        Scot Armstrong   
999  A stuffy businessman finds himself trapped ins...      Barry Sonnenfeld   

                                                Actors  Year  \
0    Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014   
1    Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012   
2    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016   
3    Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016   
4    Will Smith, Jared Leto, Margot Robbie, Viola D...  2016   
..                                                 ...   ...   
995  Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...  2015   
996  Lauren German, Heather Matarazzo, Bijou Philli...  2007   
997  Robert Hoffman, Briana Evigan, Cassie Ventura,...  2008   
998  Adam Pally, T.J. Miller, Thomas Middleditch,Sh...  2014   
999  Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...  2016   

     Runtime (Minutes)  Rating   Votes  Revenue (Millions)  Metascore  
0                  121     8.1  757074              333.13       76.0  
1                  124     7.0  485820              126.46       65.0  
2                  117     7.3  157606              138.12       62.0  
3                  108     7.2   60545              270.32       59.0  
4                  123     6.2  393727              325.02       40.0  
..                 ...     ...     ...                 ...        ...  
995                111     6.2   27585                 NaN       45.0  
996                 94     5.5   73152               17.54       46.0  
997                 98     6.2   70699               58.01       50.0  
998                 93     5.6    4881                 NaN       22.0  
999                 87     5.3   12435               19.64       11.0  

[1000 rows x 12 columns]

We're loading this dataset from a CSV and designating the movie titles to be our index.

# What is Index? How Row and Column Index is differ?

Index is like an address, that’s how any data point across the dataframe or series can be accessed. 

Rows and columns both have indexes, rows indices are called as index and for columns its general column names.

By default row has integer index start from 0. if we want first column as index , we use "index_col" attribute and give the name of column from dataset which is used as index.

In [73]:
print("Row index is :\n\n{}".format(movie.index))
print("\n\nColumn index is :\n\n{}".format(movie.columns))

Row index is :

RangeIndex(start=0, stop=1000, step=1)


Column index is :

Index(['Rank', 'Title', 'Genre', 'Description', 'Director', 'Actors', 'Year',
       'Runtime (Minutes)', 'Rating', 'Votes', 'Revenue (Millions)',
       'Metascore'],
      dtype='object')


In [74]:
# Here we change by default row index and make first column "Title" as index

In [6]:
movies_df = pd.read_csv("C:\Python310\\imdbmovie.csv", index_col="Title")

In [75]:
print("Row index is :\n\n{}".format(movies_df.index))
print("\n\nColumn index is :\n\n{}".format(movies_df.columns))

Row index is :

Index(['Guardians of the Galaxy', 'Prometheus', 'Split', 'Sing',
       'Suicide Squad', 'The Great Wall', 'La La Land', 'Mindhorn',
       'The Lost City of Z', 'Passengers',
       ...
       'Underworld: Rise of the Lycans', 'Taare Zameen Par',
       'Take Me Home Tonight', 'Resident Evil: Afterlife', 'Project X',
       'Secret in Their Eyes', 'Hostel: Part II', 'Step Up 2: The Streets',
       'Search Party', 'Nine Lives'],
      dtype='object', name='Title', length=1000)


Column index is :

Index(['Rank', 'Genre', 'Description', 'Director', 'Actors', 'Year',
       'Runtime (Minutes)', 'Rating', 'Votes', 'Revenue (Millions)',
       'Metascore'],
      dtype='object')


In [7]:
movies_df

Rank                     Genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   
Split                       3           Horror,Thriller   
Sing                        4   Animation,Comedy,Family   
Suicide Squad               5  Action,Adventure,Fantasy   
...                       ...                       ...   
Secret in Their Eyes      996       Crime,Drama,Mystery   
Hostel: Part II           997                    Horror   
Step Up 2: The Streets    998       Drama,Music,Romance   
Search Party              999          Adventure,Comedy   
Nine Lives               1000     Comedy,Family,Fantasy   

                                                               Description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   
...                                                                    ...   
Secret in Their Eyes     A tight-knit team of rising investigators, alo...   
Hostel: Part II          Three American college students studying abroa...   
Step Up 2: The Streets   Romantic sparks occur between two dance studen...   
Search Party             A pair of friends embark on a mission to reuni...   
Nine Lives               A stuffy businessman finds himself trapped ins...   

                                     Director  \
Title                                           
Guardians of the Galaxy            James Gunn   
Prometheus                       Ridley Scott   
Split                      M. Night Shyamalan   
Sing                     Christophe Lourdelet   
Suicide Squad                      David Ayer   
...                                       ...   
Secret in Their Eyes                Billy Ray   
Hostel: Part II                      Eli Roth   
Step Up 2: The Streets             Jon M. Chu   
Search Party                   Scot Armstrong   
Nine Lives                   Barry Sonnenfeld   

                                                                    Actors  \
Title                                                                        
Guardians of the Galaxy  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus               Noomi Rapace, Logan Marshall-Green, Michael Fa...   
Split                    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...   
Sing                     Matthew McConaughey,Reese Witherspoon, Seth Ma...   
Suicide Squad            Will Smith, Jared Leto, Margot Robbie, Viola D...   
...                                                                    ...   
Secret in Their Eyes     Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...   
Hostel: Part II          Lauren German, Heather Matarazzo, Bijou Philli...   
Step Up 2: The Streets   Robert Hoffman, Briana Evigan, Cassie Ventura,...   
Search Party             Adam Pally, T.J. Miller, Thomas Middleditch,Sh...   
Nine Lives               Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...   

                         Year  Runtime (Minutes)  Rating   Votes  \
Title                                                              
Guardians of the Galaxy  2014                121     8.1  757074   
Prometheus               2012                124     7.0  485820   
Split                    2016                117     7.3  157606   
Sing                     2016                108     7.2   60545   
Suicide Squad            2016                123     6.2  393727   
...                       ...                ...     ...     ...   
Secret in Their Eyes     2015                

See the difference between the output of both dataframe. 

movie dataframe contain row index as 0,1,2,.... means default way 

movies_df dataframe set title as row index so we can access row based on giving movies title.

Both output display all records of the file as well as last row show the total row & column in file (1000 row and 11 columns)

# Display information about File

.info() should be one of the very first commands you run after loading your data:

In [15]:
movies_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1000 entries, Guardians of the Galaxy to Nine Lives
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Rank                1000 non-null   int64  
 1   Genre               1000 non-null   object 
 2   Description         1000 non-null   object 
 3   Director            1000 non-null   object 
 4   Actors              1000 non-null   object 
 5   Year                1000 non-null   int64  
 6   Runtime (Minutes)   1000 non-null   int64  
 7   Rating              1000 non-null   float64
 8   Votes               1000 non-null   int64  
 9   Revenue (Millions)  872 non-null    float64
 10  Metascore           936 non-null    float64
dtypes: float64(3), int64(4), object(4)
memory usage: 93.8+ KB


**<u>info()</u>**: Provides the essential details about your dataset, such as **the number of rows and columns, the number of non-null values, what type of data is in each column, and how much memory your DataFrame is using**.

Notice in our movies dataset **we have some obvious missing values in the Revenue and Metascore columns - check Non Null vlaue of all column you can get your answer how to find?**. We'll look at how to handle those in a bit.

Seeing the datatype quickly is actually quite useful. Sometime you just imported some JSON or other file and the integers were recorded as strings. You go to do some arithmetic and find an "unsupported operand" Exception because you can't do math with strings. Calling .info() will quickly point out that your column you thought was all integers are actually string objects.

Another fast and useful attribute is .shape, which outputs just a tuple of (rows, columns):

In [17]:
movies_df.shape

(1000, 11)

Note that **.shape** has <u>no parentheses</u> and is a simple **tuple of format (rows, columns). So we have 1000 rows and 11 columns in our movies DataFrame**.

shape is useful in some situation like

when cleaning and transforming data. For example, you might filter some rows based on some criteria and then want to know quickly how many rows were removed.

# Viewing Data

viewing first few row use head() 

In [10]:
movies_df.head()

Rank                     Genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   
Split                       3           Horror,Thriller   
Sing                        4   Animation,Comedy,Family   
Suicide Squad               5  Action,Adventure,Fantasy   

                                                               Description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   

                                     Director  \
Title                                           
Guardians of the Galaxy            James Gunn   
Prometheus                       Ridley Scott   
Split                      M. Night Shyamalan   
Sing                     Christophe Lourdelet   
Suicide Squad                      David Ayer   

                                                                    Actors  \
Title                                                                        
Guardians of the Galaxy  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus               Noomi Rapace, Logan Marshall-Green, Michael Fa...   
Split                    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...   
Sing                     Matthew McConaughey,Reese Witherspoon, Seth Ma...   
Suicide Squad            Will Smith, Jared Leto, Margot Robbie, Viola D...   

                         Year  Runtime (Minutes)  Rating   Votes  \
Title                                                              
Guardians of the Galaxy  2014                121     8.1  757074   
Prometheus               2012                124     7.0  485820   
Split                    2016                117     7.3  157606   
Sing                     2016                108     7.2   60545   
Suicide Squad            2016                123     6.2  393727   

                         Revenue (Millions)  Metascore  
Title                                                   
Guardians of the Galaxy              333.13       76.0  
Prometheus                           126.46       65.0  
Split                                138.12       62.0  
Sing                                 270.32       59.0  
Suicide Squad                        325.02       40.0


.head() outputs the first five rows of your DataFrame by default, but we could also pass a number as well: movies_df.head(10) would output the top ten rows, for example.



In [11]:
movies_df.head(-100)

Rank                     Genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   
Split                       3           Horror,Thriller   
Sing                        4   Animation,Comedy,Family   
Suicide Squad               5  Action,Adventure,Fantasy   
...                       ...                       ...   
Planet Terror             896      Action,Comedy,Horror   
Concussion                897     Biography,Drama,Sport   
The Fall                  898    Adventure,Comedy,Drama   
The Ugly Truth            899            Comedy,Romance   
Bride Wars                900            Comedy,Romance   

                                                               Description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   
...                                                                    ...   
Planet Terror            After an experimental bio-weapon is released, ...   
Concussion               In Pittsburgh, accomplished pathologist Dr. Be...   
The Fall                 In a hospital on the outskirts of 1920s Los An...   
The Ugly Truth           A romantically challenged morning show produce...   
Bride Wars               Two best friends become rivals when they sched...   

                                     Director  \
Title                                           
Guardians of the Galaxy            James Gunn   
Prometheus                       Ridley Scott   
Split                      M. Night Shyamalan   
Sing                     Christophe Lourdelet   
Suicide Squad                      David Ayer   
...                                       ...   
Planet Terror                Robert Rodriguez   
Concussion                    Peter Landesman   
The Fall                         Tarsem Singh   
The Ugly Truth                 Robert Luketic   
Bride Wars                        Gary Winick   

                                                                    Actors  \
Title                                                                        
Guardians of the Galaxy  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus               Noomi Rapace, Logan Marshall-Green, Michael Fa...   
Split                    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...   
Sing                     Matthew McConaughey,Reese Witherspoon, Seth Ma...   
Suicide Squad            Will Smith, Jared Leto, Margot Robbie, Viola D...   
...                                                                    ...   
Planet Terror            Rose McGowan, Freddy Rodríguez, Josh Brolin,Ma...   
Concussion               Will Smith, Alec Baldwin, Albert Brooks, David...   
The Fall                 Lee Pace, Catinca Untaru, Justine Waddell, Kim...   
The Ugly Truth           Katherine Heigl, Gerard Butler, Bree Turner, E...   
Bride Wars               Kate Hudson, Anne Hathaway, Candice Bergen, Br...   

                         Year  Runtime (Minutes)  Rating   Votes  \
Title                                                              
Guardians of the Galaxy  2014                121     8.1  757074   
Prometheus               2012                124     7.0  485820   
Split                    2016                117     7.3  157606   
Sing                     2016                108     7.2   60545   
Suicide Squad            2016                123     6.2  393727   
...                       ...                ...     ...     ...   
Planet Terror            2007                

To see the last five rows use .tail(). tail() also accepts a number, and in this case we printing the bottom two rows.:

In [12]:
movies_df.tail()

Rank                  Genre  \
Title                                                 
Secret in Their Eyes     996    Crime,Drama,Mystery   
Hostel: Part II          997                 Horror   
Step Up 2: The Streets   998    Drama,Music,Romance   
Search Party             999       Adventure,Comedy   
Nine Lives              1000  Comedy,Family,Fantasy   

                                                              Description  \
Title                                                                       
Secret in Their Eyes    A tight-knit team of rising investigators, alo...   
Hostel: Part II         Three American college students studying abroa...   
Step Up 2: The Streets  Romantic sparks occur between two dance studen...   
Search Party            A pair of friends embark on a mission to reuni...   
Nine Lives              A stuffy businessman finds himself trapped ins...   

                                Director  \
Title                                      
Secret in Their Eyes           Billy Ray   
Hostel: Part II                 Eli Roth   
Step Up 2: The Streets        Jon M. Chu   
Search Party              Scot Armstrong   
Nine Lives              Barry Sonnenfeld   

                                                                   Actors  \
Title                                                                       
Secret in Their Eyes    Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...   
Hostel: Part II         Lauren German, Heather Matarazzo, Bijou Philli...   
Step Up 2: The Streets  Robert Hoffman, Briana Evigan, Cassie Ventura,...   
Search Party            Adam Pally, T.J. Miller, Thomas Middleditch,Sh...   
Nine Lives              Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...   

                        Year  Runtime (Minutes)  Rating  Votes  \
Title                                                            
Secret in Their Eyes    2015                111     6.2  27585   
Hostel: Part II         2007                 94     5.5  73152   
Step Up 2: The Streets  2008                 98     6.2  70699   
Search Party            2014                 93     5.6   4881   
Nine Lives              2016                 87     5.3  12435   

                        Revenue (Millions)  Metascore  
Title                                                  
Secret in Their Eyes                   NaN       45.0  
Hostel: Part II                      17.54       46.0  
Step Up 2: The Streets               58.01       50.0  
Search Party                           NaN       22.0  
Nine Lives                           19.64       11.0

In [13]:
movies_df.tail(-100)

Rank                     Genre  \
Title                                                     
Legend                    101     Biography,Crime,Drama   
Thor                      102  Action,Adventure,Fantasy   
The Martian               103    Adventure,Drama,Sci-Fi   
Contratiempo              104    Crime,Mystery,Thriller   
The Man from U.N.C.L.E.   105   Action,Adventure,Comedy   
...                       ...                       ...   
Secret in Their Eyes      996       Crime,Drama,Mystery   
Hostel: Part II           997                    Horror   
Step Up 2: The Streets    998       Drama,Music,Romance   
Search Party              999          Adventure,Comedy   
Nine Lives               1000     Comedy,Family,Fantasy   

                                                               Description  \
Title                                                                        
Legend                   Identical twin gangsters Ronald and Reginald K...   
Thor                     The powerful but arrogant god Thor is cast out...   
The Martian              An astronaut becomes stranded on Mars after hi...   
Contratiempo             A young businessman faces a lawyer trying to p...   
The Man from U.N.C.L.E.  In the early 1960s, CIA agent Napoleon Solo an...   
...                                                                    ...   
Secret in Their Eyes     A tight-knit team of rising investigators, alo...   
Hostel: Part II          Three American college students studying abroa...   
Step Up 2: The Streets   Romantic sparks occur between two dance studen...   
Search Party             A pair of friends embark on a mission to reuni...   
Nine Lives               A stuffy businessman finds himself trapped ins...   

                                 Director  \
Title                                       
Legend                    Brian Helgeland   
Thor                      Kenneth Branagh   
The Martian                  Ridley Scott   
Contratiempo                  Oriol Paulo   
The Man from U.N.C.L.E.       Guy Ritchie   
...                                   ...   
Secret in Their Eyes            Billy Ray   
Hostel: Part II                  Eli Roth   
Step Up 2: The Streets         Jon M. Chu   
Search Party               Scot Armstrong   
Nine Lives               Barry Sonnenfeld   

                                                                    Actors  \
Title                                                                        
Legend                   Tom Hardy, Emily Browning, Taron Egerton, Paul...   
Thor                     Chris Hemsworth, Anthony Hopkins, Natalie Port...   
The Martian              Matt Damon, Jessica Chastain, Kristen Wiig, Ka...   
Contratiempo             Mario Casas, Ana Wagener, José Coronado, Bárba...   
The Man from U.N.C.L.E.  Henry Cavill, Armie Hammer, Alicia Vikander, E...   
...                                                                    ...   
Secret in Their Eyes     Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...   
Hostel: Part II          Lauren German, Heather Matarazzo, Bijou Philli...   
Step Up 2: The Streets   Robert Hoffman, Briana Evigan, Cassie Ventura,...   
Search Party             Adam Pally, T.J. Miller, Thomas Middleditch,Sh...   
Nine Lives               Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...   

                         Year  Runtime (Minutes)  Rating   Votes  \
Title                                                              
Legend                   2015                132     7.0  108836   
Thor                     2011                115     7.0  570814   
The Martian              2015                144     8.0  556097   
Contratiempo             2016                106     7.9    7204   
The Man from U.N.C.L.E.  2015                116     7.3  202973   
...                       ...                ...     ...     ...   
Secret in Their Eyes     2015                111     6.2   27585   
Hostel: Part II          2007

# Handling Duplicate value

**<u>drop_duplicates()</u>** : This method is used to drop any duplicate row

In [24]:
print("Shape of movies_df is {}".format(movies_df.shape))

temp_df = movies_df.drop_duplicates()

print("Shape of movies_df is {}".format(temp_df.shape))

Shape of movies_df is (1000, 11)
Shape of movies_df is (1000, 11)


Here No duplicate in dataset because after delete both dataframe shape is same .

So first we append same record in movies_df and then check.

To append row , **append()** is used.

In [25]:
temp_df = movies_df.append(movies_df) # append all row of movies_df to same dataframe so create duplicate

temp_df.shape

C:\Users\Amit\AppData\Local\Temp\ipykernel_15016\1016839374.py:1: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  temp_df = movies_df.append(movies_df) # append all row of movies_df to same dataframe so create duplicate


(2000, 11)

In [28]:
# concatenating df1 and df2 along rows

temp_df = pd.concat([movies_df,movies_df], axis=0) # append all row of movies_df to same dataframe so create duplicate

temp_df.shape



(2000, 11)

In [29]:
# To concat using columns , use axis = 1
temp = pd.concat([movies_df,movies_df], axis=1) # append all row of movies_df to same dataframe so create duplicate

temp.shape

(1000, 22)

drop_duplicates() method will also return a copy of your DataFrame, but after removing duplicates.

Calling .shape confirms we're back to the 1000 rows of our original dataset.

It's a little verbose to keep assigning DataFrames to the same variable like in this example. 

For this reason, pandas has the inplace keyword argument on many of its methods.

Using **inplace=True** will <u>modify the DataFrame object in place</u>:

In [31]:
temp_df.drop_duplicates(inplace=True)

Now our temp_df will have the transformed data automatically.

Another important argument for **drop_duplicates() is keep**, which has three possible options:

**first**: (default) Drop duplicates except for the first occurrence.
**last**: Drop duplicates except for the last occurrence.
**False**: Drop all duplicates.

Since we didn't define the keep arugment in the previous example by default it keep first record and remove other . 

This means that if two rows are the same pandas will drop the second row and keep the first row. Using last has the opposite effect: the first row is dropped.

keep, on the other hand, will drop all duplicates. If two rows are the same then both will be dropped. Watch what happens to temp_df:

In [32]:
temp_df = movies_df.append(movies_df)  # make a new copy

temp_df.drop_duplicates(inplace=True, keep=False)

temp_df.shape

C:\Users\Amit\AppData\Local\Temp\ipykernel_15016\923618369.py:1: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  temp_df = movies_df.append(movies_df)  # make a new copy


(0, 11)

Since all rows were duplicates, **keep=False** <u>dropped them all row</u> resulting in zero rows being left over. 

This is required : 

one reason is that it allows you to locate all duplicates in your dataset. When conditional selections are shown below you'll see how to do that.

# Work with Missing Value

When exploring data, we most likely encounter missing or null values, which are essentially placeholders for non-existent values.

There are two options in dealing with nulls:

1 Remove row or column (completely) from the dataset.
2 Get rid of rows or columns with nulls 

Replace nulls with non-null values, a technique known as imputation


Let's calculate to total number of nulls in each column of our dataset. The first step is to check which cells in our DataFrame are null:

**isnull()** : returns a DataFrame where each cell is either True or False depending on that cell's null status.

In [33]:
movies_df.isnull()

Rank  Genre  Description  Director  Actors   Year  \
Title                                                                         
Guardians of the Galaxy  False  False        False     False   False  False   
Prometheus               False  False        False     False   False  False   
Split                    False  False        False     False   False  False   
Sing                     False  False        False     False   False  False   
Suicide Squad            False  False        False     False   False  False   
...                        ...    ...          ...       ...     ...    ...   
Secret in Their Eyes     False  False        False     False   False  False   
Hostel: Part II          False  False        False     False   False  False   
Step Up 2: The Streets   False  False        False     False   False  False   
Search Party             False  False        False     False   False  False   
Nine Lives               False  False        False     False   False  False   

                         Runtime (Minutes)  Rating  Votes  Revenue (Millions)  \
Title                                                                           
Guardians of the Galaxy              False   False  False               False   
Prometheus                           False   False  False               False   
Split                                False   False  False               False   
Sing                                 False   False  False               False   
Suicide Squad                        False   False  False               False   
...                                    ...     ...    ...                 ...   
Secret in Their Eyes                 False   False  False                True   
Hostel: Part II                      False   False  False               False   
Step Up 2: The Streets               False   False  False               False   
Search Party                         False   False  False                True   
Nine Lives                           False   False  False               False   

                         Metascore  
Title                               
Guardians of the Galaxy      False  
Prometheus                   False  
Split                        False  
Sing                         False  
Suicide Squad                False  
...                            ...  
Secret in Their Eyes         False  
Hostel: Part II              False  
Step Up 2: The Streets       False  
Search Party                 False  
Nine Lives                   False  

[1000 rows x 11 columns]

To count the number of nulls in each column we use an aggregate function for summing:

In [34]:
movies_df.isnull().sum()

Rank                    0
Genre                   0
Description             0
Director                0
Actors                  0
Year                    0
Runtime (Minutes)       0
Rating                  0
Votes                   0
Revenue (Millions)    128
Metascore              64
dtype: int64

We can see now that our data has **128** missing values for **revenue_millions** and **64** missing values for **metascore**.

# Remove null values

When we analyse data, we regularly face the dilemma of dropping or imputing null values, and is a decision that requires intimate knowledge of your data and its context. 

Overall, removing null data is only suggested if you have a small amount of missing data.

**.dropna()**: Remove nulls

**dropna(self, axis=0, how="any", thresh=None, subset=None, inplace=False)**

**axis:** possible values are {0 or ‘index’, 1 or ‘columns’}, default 0. If 0, drop rows with null values. If 1, drop columns with missing values.

**how:** possible values are {‘any’, ‘all’}, default ‘any’. If ‘any’, drop the row/column if any of the values is null. If ‘all’, drop the row/column if all the values are missing.

**thresh:** an int value to specify the threshold for the drop operation.

**subset:** specifies the rows/columns to look for null values.

**inplace:** a boolean value. If True, the source DataFrame is changed and None is returned.



In [35]:
movies_df.dropna()

Rank                     Genre  \
Title                                                      
Guardians of the Galaxy      1   Action,Adventure,Sci-Fi   
Prometheus                   2  Adventure,Mystery,Sci-Fi   
Split                        3           Horror,Thriller   
Sing                         4   Animation,Comedy,Family   
Suicide Squad                5  Action,Adventure,Fantasy   
...                        ...                       ...   
Resident Evil: Afterlife   994   Action,Adventure,Horror   
Project X                  995                    Comedy   
Hostel: Part II            997                    Horror   
Step Up 2: The Streets     998       Drama,Music,Romance   
Nine Lives                1000     Comedy,Family,Fantasy   

                                                                Description  \
Title                                                                         
Guardians of the Galaxy   A group of intergalactic criminals are forced ...   
Prometheus                Following clues to the origin of mankind, a te...   
Split                     Three girls are kidnapped by a man with a diag...   
Sing                      In a city of humanoid animals, a hustling thea...   
Suicide Squad             A secret government agency recruits some of th...   
...                                                                     ...   
Resident Evil: Afterlife  While still out to destroy the evil Umbrella C...   
Project X                 3 high school seniors throw a birthday party t...   
Hostel: Part II           Three American college students studying abroa...   
Step Up 2: The Streets    Romantic sparks occur between two dance studen...   
Nine Lives                A stuffy businessman finds himself trapped ins...   

                                      Director  \
Title                                            
Guardians of the Galaxy             James Gunn   
Prometheus                        Ridley Scott   
Split                       M. Night Shyamalan   
Sing                      Christophe Lourdelet   
Suicide Squad                       David Ayer   
...                                        ...   
Resident Evil: Afterlife    Paul W.S. Anderson   
Project X                      Nima Nourizadeh   
Hostel: Part II                       Eli Roth   
Step Up 2: The Streets              Jon M. Chu   
Nine Lives                    Barry Sonnenfeld   

                                                                     Actors  \
Title                                                                         
Guardians of the Galaxy   Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus                Noomi Rapace, Logan Marshall-Green, Michael Fa...   
Split                     James McAvoy, Anya Taylor-Joy, Haley Lu Richar...   
Sing                      Matthew McConaughey,Reese Witherspoon, Seth Ma...   
Suicide Squad             Will Smith, Jared Leto, Margot Robbie, Viola D...   
...                                                                     ...   
Resident Evil: Afterlife  Milla Jovovich, Ali Larter, Wentworth Miller,K...   
Project X                 Thomas Mann, Oliver Cooper, Jonathan Daniel Br...   
Hostel: Part II           Lauren German, Heather Matarazzo, Bijou Philli...   
Step Up 2: The Streets    Robert Hoffman, Briana Evigan, Cassie Ventura,...   
Nine Lives                Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...   

                          Year  Runtime (Minutes)  Rating   Votes  \
Title                                                               
Guardians of the Galaxy   2014                121     8.1  757074   
Prometheus                2012                124     7.0  485820   
Split                     2016                117     7.3  157606   
Sing                      2016                108     7.2   60545   
Suicide Squad             2016                123     6.2  393727   
...                        ...                ...     .

This operation will delete any row with at least a single null value, but it will return a new DataFrame without altering the original one. Because we specify **inplace=True** in this method as well so it work on same dataframe.

So in the case of our dataset, this operation would remove 128 rows where revenue_millions is null and 64 rows where metascore is null. This obviously seems like a waste since there's perfectly good data in the other columns of those dropped rows. That's why we'll look at imputation next.

Other than just dropping rows, you can also drop columns with null values by setting axis=1:

In [36]:
movies_df.dropna(axis=1) 

Rank                     Genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   
Split                       3           Horror,Thriller   
Sing                        4   Animation,Comedy,Family   
Suicide Squad               5  Action,Adventure,Fantasy   
...                       ...                       ...   
Secret in Their Eyes      996       Crime,Drama,Mystery   
Hostel: Part II           997                    Horror   
Step Up 2: The Streets    998       Drama,Music,Romance   
Search Party              999          Adventure,Comedy   
Nine Lives               1000     Comedy,Family,Fantasy   

                                                               Description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   
...                                                                    ...   
Secret in Their Eyes     A tight-knit team of rising investigators, alo...   
Hostel: Part II          Three American college students studying abroa...   
Step Up 2: The Streets   Romantic sparks occur between two dance studen...   
Search Party             A pair of friends embark on a mission to reuni...   
Nine Lives               A stuffy businessman finds himself trapped ins...   

                                     Director  \
Title                                           
Guardians of the Galaxy            James Gunn   
Prometheus                       Ridley Scott   
Split                      M. Night Shyamalan   
Sing                     Christophe Lourdelet   
Suicide Squad                      David Ayer   
...                                       ...   
Secret in Their Eyes                Billy Ray   
Hostel: Part II                      Eli Roth   
Step Up 2: The Streets             Jon M. Chu   
Search Party                   Scot Armstrong   
Nine Lives                   Barry Sonnenfeld   

                                                                    Actors  \
Title                                                                        
Guardians of the Galaxy  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus               Noomi Rapace, Logan Marshall-Green, Michael Fa...   
Split                    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...   
Sing                     Matthew McConaughey,Reese Witherspoon, Seth Ma...   
Suicide Squad            Will Smith, Jared Leto, Margot Robbie, Viola D...   
...                                                                    ...   
Secret in Their Eyes     Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...   
Hostel: Part II          Lauren German, Heather Matarazzo, Bijou Philli...   
Step Up 2: The Streets   Robert Hoffman, Briana Evigan, Cassie Ventura,...   
Search Party             Adam Pally, T.J. Miller, Thomas Middleditch,Sh...   
Nine Lives               Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...   

                         Year  Runtime (Minutes)  Rating   Votes  
Title                                                             
Guardians of the Galaxy  2014                121     8.1  757074  
Prometheus               2012                124     7.0  485820  
Split                    2016                117     7.3  157606  
Sing                     2016                108     7.2   60545  
Suicide Squad            2016                123     6.2  393727  
...                       ...                ...     ...     ...  
Secret in Their Eyes     2015                111     

In [38]:
temp_df.shape

(0, 11)

In [39]:
temp_df.append(movies_df)

C:\Users\Amit\AppData\Local\Temp\ipykernel_15016\1824898365.py:1: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  temp_df.append(movies_df)


Rank                     Genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   
Split                       3           Horror,Thriller   
Sing                        4   Animation,Comedy,Family   
Suicide Squad               5  Action,Adventure,Fantasy   
...                       ...                       ...   
Secret in Their Eyes      996       Crime,Drama,Mystery   
Hostel: Part II           997                    Horror   
Step Up 2: The Streets    998       Drama,Music,Romance   
Search Party              999          Adventure,Comedy   
Nine Lives               1000     Comedy,Family,Fantasy   

                                                               Description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   
...                                                                    ...   
Secret in Their Eyes     A tight-knit team of rising investigators, alo...   
Hostel: Part II          Three American college students studying abroa...   
Step Up 2: The Streets   Romantic sparks occur between two dance studen...   
Search Party             A pair of friends embark on a mission to reuni...   
Nine Lives               A stuffy businessman finds himself trapped ins...   

                                     Director  \
Title                                           
Guardians of the Galaxy            James Gunn   
Prometheus                       Ridley Scott   
Split                      M. Night Shyamalan   
Sing                     Christophe Lourdelet   
Suicide Squad                      David Ayer   
...                                       ...   
Secret in Their Eyes                Billy Ray   
Hostel: Part II                      Eli Roth   
Step Up 2: The Streets             Jon M. Chu   
Search Party                   Scot Armstrong   
Nine Lives                   Barry Sonnenfeld   

                                                                    Actors  \
Title                                                                        
Guardians of the Galaxy  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus               Noomi Rapace, Logan Marshall-Green, Michael Fa...   
Split                    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...   
Sing                     Matthew McConaughey,Reese Witherspoon, Seth Ma...   
Suicide Squad            Will Smith, Jared Leto, Margot Robbie, Viola D...   
...                                                                    ...   
Secret in Their Eyes     Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...   
Hostel: Part II          Lauren German, Heather Matarazzo, Bijou Philli...   
Step Up 2: The Streets   Robert Hoffman, Briana Evigan, Cassie Ventura,...   
Search Party             Adam Pally, T.J. Miller, Thomas Middleditch,Sh...   
Nine Lives               Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...   

                         Year  Runtime (Minutes)  Rating   Votes  \
Title                                                              
Guardians of the Galaxy  2014                121     8.1  757074   
Prometheus               2012                124     7.0  485820   
Split                    2016                117     7.3  157606   
Sing                     2016                108     7.2   60545   
Suicide Squad            2016                123     6.2  393727   
...                       ...                ...     ...     ...   
Secret in Their Eyes     2015                

In [42]:
temp_df.dropna(how="all",inplace=True) # remove those row where all column contain "None" or "empty" value

In [46]:
movies_df.info()


<class 'pandas.core.frame.DataFrame'>
Index: 1000 entries, Guardians of the Galaxy to Nine Lives
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Rank                1000 non-null   int64  
 1   Genre               1000 non-null   object 
 2   Description         1000 non-null   object 
 3   Director            1000 non-null   object 
 4   Actors              1000 non-null   object 
 5   Year                1000 non-null   int64  
 6   Runtime (Minutes)   1000 non-null   int64  
 7   Rating              1000 non-null   float64
 8   Votes               1000 non-null   int64  
 9   Revenue (Millions)  872 non-null    float64
 10  Metascore           936 non-null    float64
dtypes: float64(3), int64(4), object(4)
memory usage: 93.8+ KB


In [47]:
# remove all row where revenue and metascore both column has null value

movies_df.dropna(subset=['Revenue (Millions)' and  'Metascore'])

Rank                     Genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   
Split                       3           Horror,Thriller   
Sing                        4   Animation,Comedy,Family   
Suicide Squad               5  Action,Adventure,Fantasy   
...                       ...                       ...   
Secret in Their Eyes      996       Crime,Drama,Mystery   
Hostel: Part II           997                    Horror   
Step Up 2: The Streets    998       Drama,Music,Romance   
Search Party              999          Adventure,Comedy   
Nine Lives               1000     Comedy,Family,Fantasy   

                                                               Description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   
...                                                                    ...   
Secret in Their Eyes     A tight-knit team of rising investigators, alo...   
Hostel: Part II          Three American college students studying abroa...   
Step Up 2: The Streets   Romantic sparks occur between two dance studen...   
Search Party             A pair of friends embark on a mission to reuni...   
Nine Lives               A stuffy businessman finds himself trapped ins...   

                                     Director  \
Title                                           
Guardians of the Galaxy            James Gunn   
Prometheus                       Ridley Scott   
Split                      M. Night Shyamalan   
Sing                     Christophe Lourdelet   
Suicide Squad                      David Ayer   
...                                       ...   
Secret in Their Eyes                Billy Ray   
Hostel: Part II                      Eli Roth   
Step Up 2: The Streets             Jon M. Chu   
Search Party                   Scot Armstrong   
Nine Lives                   Barry Sonnenfeld   

                                                                    Actors  \
Title                                                                        
Guardians of the Galaxy  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus               Noomi Rapace, Logan Marshall-Green, Michael Fa...   
Split                    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...   
Sing                     Matthew McConaughey,Reese Witherspoon, Seth Ma...   
Suicide Squad            Will Smith, Jared Leto, Margot Robbie, Viola D...   
...                                                                    ...   
Secret in Their Eyes     Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...   
Hostel: Part II          Lauren German, Heather Matarazzo, Bijou Philli...   
Step Up 2: The Streets   Robert Hoffman, Briana Evigan, Cassie Ventura,...   
Search Party             Adam Pally, T.J. Miller, Thomas Middleditch,Sh...   
Nine Lives               Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...   

                         Year  Runtime (Minutes)  Rating   Votes  \
Title                                                              
Guardians of the Galaxy  2014                121     8.1  757074   
Prometheus               2012                124     7.0  485820   
Split                    2016                117     7.3  157606   
Sing                     2016                108     7.2   60545   
Suicide Squad            2016                123     6.2  393727   
...                       ...                ...     ...     ...   
Secret in Their Eyes     2015                

# Imputation

Imputation is a technique used to keep valuable data that have null values.

There may be instances where dropping every row with a null value removes too big a chunk from your dataset, **so instead we can impute that null with another value, usually <u>the mean or the median</u> of that column**.

Let's look at imputing the missing values in the revenue_millions column. First we'll extract that column into its own variable:

In [50]:
revenue = movies_df['Revenue (Millions)']
revenue

Title
Guardians of the Galaxy    333.13
Prometheus                 126.46
Split                      138.12
Sing                       270.32
Suicide Squad              325.02
                            ...  
Secret in Their Eyes          NaN
Hostel: Part II             17.54
Step Up 2: The Streets      58.01
Search Party                  NaN
Nine Lives                  19.64
Name: Revenue (Millions), Length: 1000, dtype: float64

In [51]:
revenue_mean = revenue.mean()  # calculate mean of this field so we replace it in place of null

revenue_mean

82.95637614678898

With the mean, let's fill the nulls using **fillna()**

In [52]:
revenue.fillna(revenue_mean, inplace=True)

We have now replaced all nulls in revenue with the mean of the column. Notice that by using inplace=True we have actually affected the original movies_df:

In [53]:
movies_df.isnull().sum()

Rank                   0
Genre                  0
Description            0
Director               0
Actors                 0
Year                   0
Runtime (Minutes)      0
Rating                 0
Votes                  0
Revenue (Millions)     0
Metascore             64
dtype: int64

# Understanding Variables

**describe()**:  Get a summary of the distribution of continuous (numeric) variables:

In [55]:
movies_df.describe()

Rank         Year  Runtime (Minutes)       Rating         Votes  \
count  1000.000000  1000.000000        1000.000000  1000.000000  1.000000e+03   
mean    500.500000  2012.783000         113.172000     6.723200  1.698083e+05   
std     288.819436     3.205962          18.810908     0.945429  1.887626e+05   
min       1.000000  2006.000000          66.000000     1.900000  6.100000e+01   
25%     250.750000  2010.000000         100.000000     6.200000  3.630900e+04   
50%     500.500000  2014.000000         111.000000     6.800000  1.107990e+05   
75%     750.250000  2016.000000         123.000000     7.400000  2.399098e+05   
max    1000.000000  2016.000000         191.000000     9.000000  1.791916e+06   

       Revenue (Millions)   Metascore  
count         1000.000000  936.000000  
mean            82.956376   58.985043  
std             96.412043   17.194757  
min              0.000000   11.000000  
25%             17.442500   47.000000  
50%             60.375000   59.500000  
75%             99.177500   72.000000  
max            936.630000  100.000000

describe() can also be used on a categorical variable to get the count of rows, unique count of categories, top category, and freq of top category:

In [57]:
movies_df['Genre'].describe()

count                        1000
unique                        207
top       Action,Adventure,Sci-Fi
freq                           50
Name: Genre, dtype: object

This tells us that the genre column has 207 unique values, the top value is Action/Adventure/Sci-Fi, which shows up 50 times (freq).

<u>**How to know how many time(Frequency) other movie category comes?**</u>

**.value_counts()** can tell us the frequency of all values in a columns

In [59]:
movies_df['Genre'].value_counts().head(10)

Action,Adventure,Sci-Fi       50
Drama                         48
Comedy,Drama,Romance          35
Comedy                        32
Drama,Romance                 31
Animation,Adventure,Comedy    27
Action,Adventure,Fantasy      27
Comedy,Drama                  27
Comedy,Romance                26
Crime,Drama,Thriller          24
Name: Genre, dtype: int64

# DataFrame slicing, selecting, extracting

### By Column

In [62]:
genre_col = movies_df['Genre']

type(genre_col)

pandas.core.series.Series

This will return a Series. To extract a column as a DataFrame, you need to pass a list of column names. In our case that's just a single column:

In [64]:
genre_col = movies_df[['Genre']]

type(genre_col)


pandas.core.frame.DataFrame

In [66]:
subset = movies_df[['Genre', 'Rating']]

subset.head()

Genre  Rating
Title                                                    
Guardians of the Galaxy   Action,Adventure,Sci-Fi     8.1
Prometheus               Adventure,Mystery,Sci-Fi     7.0
Split                             Horror,Thriller     7.3
Sing                      Animation,Comedy,Family     7.2
Suicide Squad            Action,Adventure,Fantasy     6.2

### By Row

For rows, we have two options:

**.loc** - locates by name

**.iloc**- locates by numerical index

In [89]:
prom = movies_df.loc["Prometheus"]

prom

Rank                                                                  2
Genre                                          Adventure,Mystery,Sci-Fi
Description           Following clues to the origin of mankind, a te...
Director                                                   Ridley Scott
Actors                Noomi Rapace, Logan Marshall-Green, Michael Fa...
Year                                                               2012
Runtime (Minutes)                                                   124
Rating                                                              7.0
Votes                                                            485820
Revenue (Millions)                                               126.46
Metascore                                                          65.0
Name: Prometheus, dtype: object

In [90]:
prom = movies_df.iloc[1]
prom

Rank                                                                  2
Genre                                          Adventure,Mystery,Sci-Fi
Description           Following clues to the origin of mankind, a te...
Director                                                   Ridley Scott
Actors                Noomi Rapace, Logan Marshall-Green, Michael Fa...
Year                                                               2012
Runtime (Minutes)                                                   124
Rating                                                              7.0
Votes                                                            485820
Revenue (Millions)                                               126.46
Metascore                                                          65.0
Name: Prometheus, dtype: object

In [92]:
movie_subset = movies_df.loc['Prometheus':'Sing']

movie_subset

Rank                     Genre  \
Title                                        
Prometheus     2  Adventure,Mystery,Sci-Fi   
Split          3           Horror,Thriller   
Sing           4   Animation,Comedy,Family   

                                                  Description  \
Title                                                           
Prometheus  Following clues to the origin of mankind, a te...   
Split       Three girls are kidnapped by a man with a diag...   
Sing        In a city of humanoid animals, a hustling thea...   

                        Director  \
Title                              
Prometheus          Ridley Scott   
Split         M. Night Shyamalan   
Sing        Christophe Lourdelet   

                                                       Actors  Year  \
Title                                                                 
Prometheus  Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012   
Split       James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016   
Sing        Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016   

            Runtime (Minutes)  Rating   Votes  Revenue (Millions)  Metascore  
Title                                                                         
Prometheus                124     7.0  485820              126.46       65.0  
Split                     117     7.3  157606              138.12       62.0  
Sing                      108     7.2   60545              270.32       59.0

In [95]:
movie_subset = movies_df.iloc[1:4]
movie_subset

Rank                     Genre  \
Title                                        
Prometheus     2  Adventure,Mystery,Sci-Fi   
Split          3           Horror,Thriller   
Sing           4   Animation,Comedy,Family   

                                                  Description  \
Title                                                           
Prometheus  Following clues to the origin of mankind, a te...   
Split       Three girls are kidnapped by a man with a diag...   
Sing        In a city of humanoid animals, a hustling thea...   

                        Director  \
Title                              
Prometheus          Ridley Scott   
Split         M. Night Shyamalan   
Sing        Christophe Lourdelet   

                                                       Actors  Year  \
Title                                                                 
Prometheus  Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012   
Split       James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016   
Sing        Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016   

            Runtime (Minutes)  Rating   Votes  Revenue (Millions)  Metascore  
Title                                                                         
Prometheus                124     7.0  485820              126.46       65.0  
Split                     117     7.3  157606              138.12       62.0  
Sing                      108     7.2   60545              270.32       59.0

In [98]:
movie_subset = movies_df.iloc[1:2,3:4]
movie_subset

Director
Title                   
Prometheus  Ridley Scott

### Conditional Selection

We’ve gone over how to select columns and rows, but what if we want to make a conditional selection?

For example, what if we want to filter our movies DataFrame to show only films directed by Ridley Scott or films with a rating greater than or equal to 8.0?

To do that, we take a column from the DataFrame and apply a Boolean condition to it. Here's an example of a Boolean condition:

In [101]:
condition = (movies_df['Director'] == "Ridley Scott")

condition.head()

Title
Guardians of the Galaxy    False
Prometheus                  True
Split                      False
Sing                       False
Suicide Squad              False
Name: Director, dtype: bool

In [102]:
print(movies_df['Director'] == "Ridley Scott")

Title
Guardians of the Galaxy    False
Prometheus                  True
Split                      False
Sing                       False
Suicide Squad              False
                           ...  
Secret in Their Eyes       False
Hostel: Part II            False
Step Up 2: The Streets     False
Search Party               False
Nine Lives                 False
Name: Director, Length: 1000, dtype: bool


Similar to isnull(), **this returns a Series of True and False values**: True for films directed by Ridley Scott and False for ones not directed by him.

**We want to filter out all movies not directed by Ridley Scott, in other words, we don’t want the False films.**

To return the rows where that condition is True we have to pass this operation into the DataFrame:

In [104]:
movies_df[movies_df['Director'] == "Ridley Scott"].head()

Rank                     Genre  \
Title                                                    
Prometheus                 2  Adventure,Mystery,Sci-Fi   
The Martian              103    Adventure,Drama,Sci-Fi   
Robin Hood               388    Action,Adventure,Drama   
American Gangster        471     Biography,Crime,Drama   
Exodus: Gods and Kings   517    Action,Adventure,Drama   

                                                              Description  \
Title                                                                       
Prometheus              Following clues to the origin of mankind, a te...   
The Martian             An astronaut becomes stranded on Mars after hi...   
Robin Hood              In 12th century England, Robin and his band of...   
American Gangster       In 1970s America, a detective works to bring d...   
Exodus: Gods and Kings  The defiant leader Moses rises up against the ...   

                            Director  \
Title                                  
Prometheus              Ridley Scott   
The Martian             Ridley Scott   
Robin Hood              Ridley Scott   
American Gangster       Ridley Scott   
Exodus: Gods and Kings  Ridley Scott   

                                                                   Actors  \
Title                                                                       
Prometheus              Noomi Rapace, Logan Marshall-Green, Michael Fa...   
The Martian             Matt Damon, Jessica Chastain, Kristen Wiig, Ka...   
Robin Hood              Russell Crowe, Cate Blanchett, Matthew Macfady...   
American Gangster       Denzel Washington, Russell Crowe, Chiwetel Eji...   
Exodus: Gods and Kings  Christian Bale, Joel Edgerton, Ben Kingsley, S...   

                        Year  Runtime (Minutes)  Rating   Votes  \
Title                                                             
Prometheus              2012                124     7.0  485820   
The Martian             2015                144     8.0  556097   
Robin Hood              2010                140     6.7  221117   
American Gangster       2007                157     7.8  337835   
Exodus: Gods and Kings  2014                150     6.0  137299   

                        Revenue (Millions)  Metascore  
Title                                                  
Prometheus                          126.46       65.0  
The Martian                         228.43       80.0  
Robin Hood                          105.22       53.0  
American Gangster                   130.13       76.0  
Exodus: Gods and Kings               65.01       52.0

We can get used to looking at these conditionals by reading it like:

   **Select movies_df where movies_df director equals Ridley Scott**

Let's look at conditional selections using numerical values by filtering the DataFrame by ratings:

In [107]:
movies_df[movies_df['Rating'] >= 8.6].head(3)


Rank                    Genre  \
Title                                            
Interstellar       37   Adventure,Drama,Sci-Fi   
The Dark Knight    55       Action,Crime,Drama   
Inception          81  Action,Adventure,Sci-Fi   

                                                       Description  \
Title                                                                
Interstellar     A team of explorers travel through a wormhole ...   
The Dark Knight  When the menace known as the Joker wreaks havo...   
Inception        A thief, who steals corporate secrets through ...   

                          Director  \
Title                                
Interstellar     Christopher Nolan   
The Dark Knight  Christopher Nolan   
Inception        Christopher Nolan   

                                                            Actors  Year  \
Title                                                                      
Interstellar     Matthew McConaughey, Anne Hathaway, Jessica Ch...  2014   
The Dark Knight  Christian Bale, Heath Ledger, Aaron Eckhart,Mi...  2008   
Inception        Leonardo DiCaprio, Joseph Gordon-Levitt, Ellen...  2010   

                 Runtime (Minutes)  Rating    Votes  Revenue (Millions)  \
Title                                                                     
Interstellar                   169     8.6  1047747              187.99   
The Dark Knight                152     9.0  1791916              533.32   
Inception                      148     8.8  1583625              292.57   

                 Metascore  
Title                       
Interstellar          74.0  
The Dark Knight       82.0  
Inception             74.0

We can make some richer conditionals by using **logical operators | for "or" and & for "and"**.

Let's filter the the DataFrame to show only movies by Christopher Nolan OR Ridley Scott:

By using the correlation method **.corr()** we can generate the relationship between each **continuous** variable:

In [109]:
movies_df[(movies_df['Director'] == 'Christopher Nolan') | (movies_df['Director'] == 'Ridley Scott')].head()

Rank                     Genre  \
Title                                             
Prometheus          2  Adventure,Mystery,Sci-Fi   
Interstellar       37    Adventure,Drama,Sci-Fi   
The Dark Knight    55        Action,Crime,Drama   
The Prestige       65      Drama,Mystery,Sci-Fi   
Inception          81   Action,Adventure,Sci-Fi   

                                                       Description  \
Title                                                                
Prometheus       Following clues to the origin of mankind, a te...   
Interstellar     A team of explorers travel through a wormhole ...   
The Dark Knight  When the menace known as the Joker wreaks havo...   
The Prestige     Two stage magicians engage in competitive one-...   
Inception        A thief, who steals corporate secrets through ...   

                          Director  \
Title                                
Prometheus            Ridley Scott   
Interstellar     Christopher Nolan   
The Dark Knight  Christopher Nolan   
The Prestige     Christopher Nolan   
Inception        Christopher Nolan   

                                                            Actors  Year  \
Title                                                                      
Prometheus       Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012   
Interstellar     Matthew McConaughey, Anne Hathaway, Jessica Ch...  2014   
The Dark Knight  Christian Bale, Heath Ledger, Aaron Eckhart,Mi...  2008   
The Prestige     Christian Bale, Hugh Jackman, Scarlett Johanss...  2006   
Inception        Leonardo DiCaprio, Joseph Gordon-Levitt, Ellen...  2010   

                 Runtime (Minutes)  Rating    Votes  Revenue (Millions)  \
Title                                                                     
Prometheus                     124     7.0   485820              126.46   
Interstellar                   169     8.6  1047747              187.99   
The Dark Knight                152     9.0  1791916              533.32   
The Prestige                   130     8.5   913152               53.08   
Inception                      148     8.8  1583625              292.57   

                 Metascore  
Title                       
Prometheus            65.0  
Interstellar          74.0  
The Dark Knight       82.0  
The Prestige          66.0  
Inception             74.0

We need to make sure **to group evaluations with <u>parentheses</u>** so Python knows how to evaluate the conditional.

Using the **isin()** method we could make this more concise though: It is like subquery

In [110]:
movies_df[movies_df['Director'].isin(['Christopher Nolan', 'Ridley Scott'])].head()

Rank                     Genre  \
Title                                             
Prometheus          2  Adventure,Mystery,Sci-Fi   
Interstellar       37    Adventure,Drama,Sci-Fi   
The Dark Knight    55        Action,Crime,Drama   
The Prestige       65      Drama,Mystery,Sci-Fi   
Inception          81   Action,Adventure,Sci-Fi   

                                                       Description  \
Title                                                                
Prometheus       Following clues to the origin of mankind, a te...   
Interstellar     A team of explorers travel through a wormhole ...   
The Dark Knight  When the menace known as the Joker wreaks havo...   
The Prestige     Two stage magicians engage in competitive one-...   
Inception        A thief, who steals corporate secrets through ...   

                          Director  \
Title                                
Prometheus            Ridley Scott   
Interstellar     Christopher Nolan   
The Dark Knight  Christopher Nolan   
The Prestige     Christopher Nolan   
Inception        Christopher Nolan   

                                                            Actors  Year  \
Title                                                                      
Prometheus       Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012   
Interstellar     Matthew McConaughey, Anne Hathaway, Jessica Ch...  2014   
The Dark Knight  Christian Bale, Heath Ledger, Aaron Eckhart,Mi...  2008   
The Prestige     Christian Bale, Hugh Jackman, Scarlett Johanss...  2006   
Inception        Leonardo DiCaprio, Joseph Gordon-Levitt, Ellen...  2010   

                 Runtime (Minutes)  Rating    Votes  Revenue (Millions)  \
Title                                                                     
Prometheus                     124     7.0   485820              126.46   
Interstellar                   169     8.6  1047747              187.99   
The Dark Knight                152     9.0  1791916              533.32   
The Prestige                   130     8.5   913152               53.08   
Inception                      148     8.8  1583625              292.57   

                 Metascore  
Title                       
Prometheus            65.0  
Interstellar          74.0  
The Dark Knight       82.0  
The Prestige          66.0  
Inception             74.0

### Let's say we want all movies that were released between 2005 and 2010, have a rating above 8.0, but made below the 25th percentile in revenue.



In [114]:
movies_df[
    ((movies_df['Year'] >= 2005) & (movies_df['Year'] <= 2010))
    & (movies_df['Rating'] > 8.0)
    & (movies_df['Revenue (Millions)'] < movies_df['Revenue (Millions)'].quantile(0.25))
]

Rank               Genre  \
Title                                           
3 Idiots              431        Comedy,Drama   
The Lives of Others   477      Drama,Thriller   
Incendies             714   Drama,Mystery,War   
Taare Zameen Par      992  Drama,Family,Music   

                                                           Description  \
Title                                                                    
3 Idiots             Two friends are searching for their long lost ...   
The Lives of Others  In 1984 East Berlin, an agent of the secret po...   
Incendies            Twins journey to the Middle East to discover t...   
Taare Zameen Par     An eight-year-old boy is thought to be a lazy ...   

                                             Director  \
Title                                                   
3 Idiots                              Rajkumar Hirani   
The Lives of Others  Florian Henckel von Donnersmarck   
Incendies                            Denis Villeneuve   
Taare Zameen Par                           Aamir Khan   

                                                                Actors  Year  \
Title                                                                          
3 Idiots               Aamir Khan, Madhavan, Mona Singh, Sharman Joshi  2009   
The Lives of Others  Ulrich Mühe, Martina Gedeck,Sebastian Koch, Ul...  2006   
Incendies            Lubna Azabal, Mélissa Désormeaux-Poulin, Maxim...  2010   
Taare Zameen Par     Darsheel Safary, Aamir Khan, Tanay Chheda, Sac...  2007   

                     Runtime (Minutes)  Rating   Votes  Revenue (Millions)  \
Title                                                                        
3 Idiots                           170     8.4  238789                6.52   
The Lives of Others                137     8.5  278103               11.28   
Incendies                          131     8.2   92863                6.86   
Taare Zameen Par                   165     8.5  102697                1.20   

                     Metascore  
Title                           
3 Idiots                  67.0  
The Lives of Others       89.0  
Incendies                 80.0  
Taare Zameen Par          42.0

# Find Relationship Between Variables ( Columns )

Correlation tables are a numerical representation of the bivariate relationships in the dataset.

Positive numbers indicate a positive correlation — one goes up the other goes up — and negative numbers represent an inverse correlation — one goes up the other goes down. 1.0 indicates a perfect correlation.

So looking in the first row, first column we see rank has a perfect correlation with itself, which is obvious. On the other hand, the correlation between votes and revenue_millions is 0.6. A little more interesting.

In [60]:
movies_df.corr()


Rank      Year  Runtime (Minutes)    Rating     Votes  \
Rank                1.000000 -0.261605          -0.221739 -0.219555 -0.283876   
Year               -0.261605  1.000000          -0.164900 -0.211219 -0.411904   
Runtime (Minutes)  -0.221739 -0.164900           1.000000  0.392214  0.407062   
Rating             -0.219555 -0.211219           0.392214  1.000000  0.511537   
Votes              -0.283876 -0.411904           0.407062  0.511537  1.000000   
Revenue (Millions) -0.252996 -0.117562           0.247834  0.189527  0.607941   
Metascore          -0.191869 -0.079305           0.211978  0.631897  0.325684   

                    Revenue (Millions)  Metascore  
Rank                         -0.252996  -0.191869  
Year                         -0.117562  -0.079305  
Runtime (Minutes)             0.247834   0.211978  
Rating                        0.189527   0.631897  
Votes                         0.607941   0.325684  
Revenue (Millions)            1.000000   0.133328  
Metascore                     0.133328   1.000000